src/tiled/terraindock.cpp

```diff
@@ -214,6 +227,13 @@ void TerrainDock::setDocument(Document *document)
         connect(terrainModel, &TilesetTerrainModel::terrainRemoved,
                 this, &TerrainDock::refreshCurrentTerrain);
 
+        /*
+         * The current terrain does not change when moving terrains.
+         * We need to refresh this in order to disable the up/down buttons when
+         * appropriate.
+         */
+        connect(terrainModel, &TilesetTerrainModel::updateTerrainDock,
```

Comment:
I think you could just use `QAbstractItemModel::rowsMoved` and drop the special `updateTerrainDock` signal.

Reply:
I tried that, but it gets triggered when `beginMoveRows()` executes, despite what the QT docs imply. We need a signal that happens after `endMoveRows()`, so I made `updateTerrainDock()`.

Reply:
Okay, I may have been wrong here. I think there is a bug in my code, as `QAbstractItemModel::rowsMoved` should really work.

Reply:
Ah, I was passing a TerrainModel pointer instead of TilesetTerrainModel to `connect()`. It works perfectly with the `QAbstractItemModel::rowsMoved` signal. I'll update my PR.